contrib/seccomp-notify-plugin-example-rust/src/seccompwatcher.rs

```diff
@@ -0,0 +1,244 @@
+    }
+    0
+}
+
+#[no_mangle]
+pub extern "C" fn run_oci_seccomp_notify_handle_request(_opaque: *mut *mut core::ffi::c_void,
+                                                        _sizes: *mut nc::seccomp_notif_sizes_t,
+                                                        sreq: *mut nc::seccomp_notif_t,
+                                                        sresp: *mut nc::seccomp_notif_resp_t,
+                                                        _seccomp_fd: i32,
+                                                        shandled: *mut i32) -> i32
+{
+    let req = unsafe {&mut *sreq};
+    let resp = unsafe {&mut *sresp};
+    let handled = unsafe {&mut *shandled};
+
+    if req.data.nr == nc::SYS_MKNOD as i32 {
+        *handled = RUN_OCI_SECCOMP_NOTIFY_HANDLE_SEND_RESPONSE;
+
+        match handle_mknod_request(req) {
+            Ok(_) => {
+                resp.error = 0;
+            }
+            Err(errno) => {
+                if errno == nc::EPERM {
+                    *handled = RUN_OCI_SECCOMP_NOTIFY_HANDLE_SEND_RESPONSE_AND_CONTINUE;
+                } else {
+                    resp.error = -errno;
+                }
+            }
+        }
+
+        resp.id = req.id;
+        resp.val = 0;
+        return 0;
+    }
+    if req.data.nr == nc::SYS_MKNODAT as i32 {
+        *handled = RUN_OCI_SECCOMP_NOTIFY_HANDLE_SEND_RESPONSE;
+
+        match handle_mknodat_request(req) {
+            Ok(_) => {
+                resp.error = 0;
+            }
+            Err(errno) => {
+                if errno == nc::EPERM {
+                    *handled = RUN_OCI_SECCOMP_NOTIFY_HANDLE_SEND_RESPONSE_AND_CONTINUE;
+                } else {
+                    resp.error = -errno;
+                }
+            }
+        }
+
+        resp.id = req.id;
+        resp.val = 0;
+        return 0;
+    }
+
+    *handled = RUN_OCI_SECCOMP_NOTIFY_HANDLE_NOT_HANDLED;
+ 0
+}
+
+fn get_string_from_buffer(fname_buffer: Vec<u8>) -> Result<String, nc::Errno> {
+    let len = match fname_buffer.iter().position(|&x| x == 0) {
+        Some(s) => {
+            s
+        }
+        None => {
+            return Err(nc::EINVAL);
+        }
+
+    };
+    let f = match CString::new(&fname_buffer[0..len]) {
+        Ok(s) => {
+            s
+        }
+        Err(_) => {
+            return Err(nc::EINVAL);
+        }
+    };
+    let filename = match f.into_string() {
+        Ok(s) => {
+            s
+        }
+        Err(_) => {
+            return Err(nc::EINVAL);
+        }
+    };
+    Ok(filename)
+}
+
+fn handle_mknodat_request(req: &mut nc::seccomp_notif_t) -> Result<u32, nc::Errno> {
+    return handle_mknod_internal(0, 1, 2, 3, req)
+}
+
+fn handle_mknod_request(req: &mut nc::seccomp_notif_t) -> Result<u32, nc::Errno> {
+    return handle_mknod_internal(-1, 0, 1, 2, req)
+
+}
+
+fn get_file_from_fd(fd: std::os::unix::io::RawFd) -> std::fs::File {
+    let r = unsafe {std::fs::File::from_raw_fd(fd)};
```

Comment:
In modules that use `File` I tend to do:
`import std::fs::File`
at the toplevel at least so that everywhere else you can just write e.g. `File::from_raw_fd()`.

Also you don't need to "name" the return value so this can just be:
```
unsafe { File::from_raw_fd(fd) }
```

But that makes it more obvious that we're just "unsafe wrapping".  I think that's ok in this instance but I'd probably wrap this up with a:
```
fn nc_open_ro(fname: &str) -> File {
  let fd = nc::open(fname, nc::O_RDONLY, 0);
  unsafe { File::from_raw_fd (fd) }
```
or so?

Reply:
Also, I think you likely want `O_CLOEXEC` everywhere.